### ISSUE TITLE
Leading zero byte removed seemingly by TEE_PopulateTransientObject

### ISSUE THREAD

**Issue by madte** [OP-TEE/optee_os#4201]
Hi there,
i am implementing testcases for ECDSA verify. All tests run well until except the P521 curve testcases. There i get the TEE_ERROR_BAD_PARAMETERS error when i call verify, which seems to be related to the lengths of TEE_ATTR_ECC_PUBLIC_VALUE_X and TEE_ATTR_ECC_PUBLIC_VALUE_Y.

Both public key coordinates are stored in uint8_t arrays with length 66. I allocate an object with type TEE_TYPE_ECDSA_PUBLIC_KEY and then add the public key coordinates and curve id using the functions TEE_InitRefAttribute and TEE_PopulateTransientObject. Here is the code snippet:

```C
uint8_t eccPkX[66] = {0};
uint8_t eccPkY[66] = {0};

DMSG("Allocating key object memory\n");
result = TEE_AllocateTransientObject(TEE_TYPE_ECDSA_PUBLIC_KEY,
	 curveLenBits, &key_handle);
if (result != TEE_SUCCESS)
{
	EMSG("Failed to alloc transient object handle : 0x%x",
		 result);
	return result;
}

DMSG("Transform hex strings to byte arrays\n");

FromHex(eccPkX, eccLen, curTestKeyPair->pkX, 1);
FromHex(eccPkY, eccLen, curTestKeyPair->pkY, 1);

DMSG("\nECC PK X (%u): ",eccLen);
printHex(eccPkX, eccLen);
DMSG("\nECC PK Y(%u): ",eccLen);
printHex(eccPkY, eccLen);

IMSG("Setting ECDSA key object parameters\n");
TEE_InitRefAttribute(&attrs[0], TEE_ATTR_ECC_PUBLIC_VALUE_X,
	 eccPkX, eccLen);
TEE_InitRefAttribute(&attrs[1], TEE_ATTR_ECC_PUBLIC_VALUE_Y,
	 eccPkY, eccLen);
TEE_InitValueAttribute(&attrs[2], TEE_ATTR_ECC_CURVE,
	   curTestKeyPair->curveId, 0);

/* Assign the attributes to the key object */
result = TEE_PopulateTransientObject(key_handle, attrs, 3);
if (result != TEE_SUCCESS)
{
	EMSG("0x%08x", result);
	return result;
}
```

Then i use the TEE_GetObjectBufferAttribute functions to determine if everything was written correctly. There i can see that the returned TEE_ATTR_ECC_PUBLIC_VALUE_X and TEE_ATTR_ECC_PUBLIC_VALUE_Y is only 65 bytes long and the leading zero byte is missing. I guess thats why TEE_AsymmetricVerifyDigest fails with TEE_ERROR_BAD_PARAMETERS. Looking at the op-tee test cases in xtest, i assume 66 bytes are the correct buffer size for P521 curve coordinates (as they are actually 65byte + 1 bit long).

Here is the output of my test program:

D/TA:  ecdsa_verify:812 Transform hex strings to byte arrays
D/TA:  ecdsa_verify:817 
**ECC PK X (66): 
00f50a08703250c15f043c8c46e99783435245cf98f4f2694b0e2f8d029a514dd6f0b086d4ed892000cd5590107aae69c4c0a7a95f7cf74e5770a07d5db55bce4ab4**
D/TA:  ecdsa_verify:819 
**ECC PK Y(66): 
00f2c770bab8b9be4cdb6ecd3dc26c698da0d2599cebf3d904f7f9ca3a55e64731810d73cd317264e50baba4bc2860857e16d6cbb79501bc9e3a32bd172ea8a71dee**
I/TA: Setting ECDSA key object parameters
D/TA:  printEccKeyObject:309 Acessing key object...
**D/TA:  printEccKeyObject:318 ECC PK X (65): 
f50a08703250c15f043c8c46e99783435245cf98f4f2694b0e2f8d029a514dd6f0b086d4ed892000cd5590107aae69c4c0a7a95f7cf74e5770a07d5db55bce4ab4
D/TA:  printEccKeyObject:327 ECC PK Y (65): 
f2c770bab8b9be4cdb6ecd3dc26c698da0d2599cebf3d904f7f9ca3a55e64731810d73cd317264e50baba4bc2860857e16d6cbb79501bc9e3a32bd172ea8a71dee**
D/TA:  printEccKeyObject:336 Curve ID: 5
D/TA:  verifyEcdsa:524 Digest (20): 
01e7cbb20c9a66abf149c79d11859051d35cfddd
D/TA:  verifyEcdsa:527 Signature (131): 
01e7cbb20c9a66abf149c79d11859051d35cfddd04f420dd23bd3206c82b29e782453cabfefe792e4e3e68c9bf6bf50d5a00ba5dd73b41378fb46e91ca797dbb2500f1e9252573c003cb77f22c8c6d56f2149f7e8d88d699983da9250c8edfd4b9f864a46c48819524651886e3fd56492f4b6c75fb50a1d59e8bfc25f9fd42dc4e1d37
D/TA:  verifyEcdsa:530 Allocating verification operation
D/TA:  prepareEcdsaOperation:365 TEE_GetObjectInfo1(): OK
D/TA:  prepareEcdsaOperation:373 TEE_AllocateOperation(): OK
D/TA:  prepareEcdsaOperation:381 TEE_SetOperationKey(): OK
I/TA: Verifying the signature...
E/TC:? 0 
E/TC:? 0 TA panicked with code 0xffff0006

I have been looking into the OP-TEE code a lot and cannot explain where the leading zero byte gets lost. Am i doing sth. wrong or could this be a bug in op-tee?

Thanks in advance for your time
Cheers 
Madte

**Comment by jenswi-linaro** [comment]
When calling `TEE_AsymmetricVerifyDigest()` you end reach `syscall_asymm_verify()` in OP-TEE Core. You could try to find where the `TEE_ERROR_BAD_PARAMETERS` error code originates from.

**Comment by madte** [comment]
Hi Jens, thanks for your answer. The error ist thrown by TEE_AsymmetricVerifyDigest(). Which makes sense as 65 bytes is too short for ECC P521 public key coordinates. My problem is:
Why does TEE_InitRefAttribute or TEE_PopulateTransientObject cut off the first zero byte in of the pkx and pky? (Bold in the output above)

**Comment by jenswi-linaro** [comment]
Some/many/most crypto algorithms requires the first bit to be set in order for it to be counted. If you think it's an error check with the specification and we'll see what can be done.

**Comment by madte** [comment]
The GPD_TEE_Internal_Core_API_Specification_v1.2.1_CC specification has no info about cutting bytes of in TEE_InitRefAttribute or TEE_PopulateTransientObject. For my tests i use NIST signature validation testcases (https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/dss/186-2ecdsatestvectors.zip
) 
For all the other curves the ECDSA tests run fine, i only have this problem with the P521 curve. This is the testcase that fails due to the first 0 being cut off in TEE_InitRefAttribute or TEE_PopulateTransientObject:

ECDSA Test P521 SHA1
Msg = 9bbbbe8a72130e1f023fb77be4648c80e1722d98bd478882383026c5c4e8748873997c5a38e0a173ed461546422d7691393dc2aceb0c0775068bc7145e33bf6a9e34f7fc6acc8f079a265168e54d3cca8d40aa04c1afd0909aa3df50908d7324aa7861b50f471fbfa5d615b0d718132c81957b178ad936deb89fde37147f8ae6
Qx = 0f50a08703250c15f043c8c46e99783435245cf98f4f2694b0e2f8d029a514dd6f0b086d4ed892000cd5590107aae69c4c0a7a95f7cf74e5770a07d5db55bce4ab4
Qy = 0f2c770bab8b9be4cdb6ecd3dc26c698da0d2599cebf3d904f7f9ca3a55e64731810d73cd317264e50baba4bc2860857e16d6cbb79501bc9e3a32bd172ea8a71dee
R = 01e7cbb20c9a66abf149c79d11859051d35cfddd04f420dd23bd3206c82b29e782453cabfefe792e4e3e68c9bf6bf50d5a00ba5dd73b41378fb46e91ca797dbb250
S = 0f1e9252573c003cb77f22c8c6d56f2149f7e8d88d699983da9250c8edfd4b9f864a46c48819524651886e3fd56492f4b6c75fb50a1d59e8bfc25f9fd42dc4e1d37
Result = P (0 )

Do you need more information of me to see what can be done?

**Comment by jforissier** [comment]
The leading zero should not matter since the keys are given in big endian format. You're supplying 66 bytes and OP-TEE converts the key into its own internal representation. When you retrieve the key with `TEE_GetObjectBufferAttribute()`, OP-TEE converts from its internal representation into a big-endian buffer again, and since the value fits in 65 bytes it is this value that you get. See https://github.com/OP-TEE/optee_os/blob/3.11.0/core/tee/tee_svc_cryp.c#L682-L683.

Of course this does not explain why your test case is failing :-/ As Jens suggested, your best bet is to trace where `TEE_ERROR_BAD_PARAMETERS` is coming from.

**Comment by madte** [comment]
Thank you jforissier for your answer. In the meanwhile i played around with the key lengths and can confirm that you are right. Supplying a 66 byte pk still gives me the same error. 

I used the symbolize script to improve the error output:

`E/TC:? 0 TA panicked with code 0xffff0006
E/LD:  Status of TA 35835a5d-24a4-4565-85d8-85e42e1bab9a
E/LD:   arch: aarch64
E/LD:  region  0: va 0x40004000 pa 0x0e32e000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x40006000 pa 0x0e330000 size 0x008000 flags r-xs (ldelf)
E/LD:  region  2: va 0x4000e000 pa 0x0e338000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x4000f000 pa 0x0e339000 size 0x004000 flags rw-s (ldelf)
E/LD:  region  4: va 0x40013000 pa 0x0e33d000 size 0x001000 flags r--s
E/LD:  region  5: va 0x40014000 pa 0x0e357000 size 0x011000 flags rw-s (stack)
E/LD:  region  6: va 0x40025000 pa 0x7dedf750 size 0x001000 flags rw-- (param)
E/LD:  region  7: va 0x40026000 pa 0x7dedf3b0 size 0x001000 flags rw-- (param)
E/LD:  region  8: va 0x40046000 pa 0x00001000 size 0x014000 flags r-xs [0] .ta_head .text .eh_frame .rodata .gnu.hash .dynsym .dynstr .hash .rela.dyn
E/LD:  region  9: va 0x4005a000 pa 0x00015000 size 0x005000 flags rw-s [0] .dynamic .got .rela.got .data .bss
E/LD:   [0] 35835a5d-24a4-4565-85d8-85e42e1bab9a @ 0x40046000 (out-br/build/optee_examples_ext-1.0/ecdsa/ta/out/35835a5d-24a4-4565-85d8-85e42e1bab9a.elf)
E/LD:  Call stack:
E/LD:   0x000000004004bff0 TEE_AsymmetricVerifyDigest at optee_os/lib/libutee/tee_api_operations.c:1761
E/LD:   0x0000000040046cc0 verifyEcdsa at out-br/build/optee_examples_ext-1.0/ecdsa/ta/ecdsa_ta.c:534
E/LD:   0x0000000040047134 ecdsa_verify at out-br/build/optee_examples_ext-1.0/ecdsa/ta/ecdsa_ta.c:845
E/LD:   0x000000004004cf60 entry_invoke_command at optee_os/lib/libutee/arch/arm/user_ta_entry.c:357
E/LD:   0x0000000040047438 __ta_entry at optee_os/out/arm/export-ta_arm64/src/user_ta_header.c:47
D/TC:? 0 user_ta_enter:167 tee_user_ta_enter: TA panicked with code 0xffff0006
D/TC:? 0 destroy_ta_ctx_from_session:313 Remove references to context (0xef4a0b0)
D/TC:? 0 destroy_context:298 Destroy TA ctx (0xef4a0b0)
`
So the error definitely originates from the TEE_AsymmetricVerifyDigest call. I will further investigate the issue and state the solution here when if figured out what the problem is.

**Comment by jforissier** [comment]
> So the error definitely originates from the TEE_AsymmetricVerifyDigest call. 

Yes, at this point it is a bit too late unfortunately so the stack is not very helpful in this case. Try adding debug traces (`DMSG("");`) in `syscall_asymm_verify()` and the function it calls, build with `TEE_CORE_LOG_LEVEL=3` and you should be able to trace down where things break.

**Comment by madte** [comment]
Sorry for the late response, almost forgot to document the solution here so it might help others in the future. Here is the whole story of the problem and how i solved it:

It was not straight forward to work with ECDSA on curve P521, TEE_AsymmetricVerifyDigest returned the BAD_PARAMETERS error code and it was difficult to find out why. As the curve size of P521 is not a multiple of 8 it was hard to determine the sizes of public key and signature (could not find any info in documentation).

My first guess was that something is wrong with the public key, as i used TEE_GetObjectBufferAttribute to print the public key x and y bytes, which only output 65 bytes each instead of the expected 66 bytes ( 521 / 8 = 65,125, therefore i assumed that at least 66 bytes are required to contain a P521 public key). After opening an issue at the OP-TEE Github, the main contributors pointed out that the TEE_GetObjectBufferAttribute does not output start bytes that are 00, therefore the public key was actually 66 bytes long but only 65 bytes were output due to the first byte being 00. With that knowledge it was likely that the BAD_PARAMETERS error is not due to the public key.

As it was probably not the public key, the problem was most likely caused by the signature. Again i could´t find any documentation about sizes, therefore i assumed it to be 131 bytes ( signature are two points R&S on curve P521, therefore 521*2/8 = 130,25, rounding up to 131 as with the public key). I tried searching the code to figure out where the BAD_PARAMETERS error originates from. It was straight forward to jump into the function TEE_AsymmetricVerifyDigest(...), but the error was not thrown within that function. To do the actual signature verification, the function _utee_asymm_verify(operation->state, ua, paramCount, digest, digestLen, signature, signatureLen)
is called. I tried to jump to the function, which lead to a function prototype in the utee_syscall.h file. That made me assume that this function call is actually a syscall where the code is not available, therefore not possible to analyze/debug where the error originates. After contact with the OP-TEE maintainers, they suggested to search for syscall_asymm_verify() and the function it calls. syscall_asymm_verify() is in the file optee_os/core/tee_svc_cryp.c and contains the syscalls that i assumed were not available as code. So there must be a mapping somewhere between _utee_asymm_verify() and syscall_asymm_verify(). syscall_asymm_verify() itself calls then crypto_acipher_ecc_verify() which is in optee_os/core/libtomcrypt/ecc.c. This finally is the actual code that does the signature verification.m Having found this code, it was possible to add debug outputs to determine where the BAD_PARAMETERS error originated from:

    /* check keysize vs sig_len */

    if ((key_size_bytes * 2) != sig_len) {

        DMSG("Key size*2: %u\n", key_size_bytes * 2);

        DMSG("sig_len: %u\n", sig_len);

        res = TEE_ERROR_BAD_PARAMETERS;

        goto out;

    }

So the error was thrown as the signature must be twice the size of a public key coordinate. Leading to 66*2 = 132. Now the only thing to figure out was the format of the signature in order to verify the test vector signatures. The test vector signatures consist of two 131 character long hexadecimal strings (R&S). A byte can be expressed as two hex chars, therefore 65,5 bytes at least required to fit R or S. As we know now that the signature is expected the be 132 bytes this won´t work without padding (→ 2*65,5 = 131 != 132). Again, could not find any information about this, so needed to try out the likely options. First try was to add a padding byte (00) before R&S giving 00||R||S. This was not correct, as the signature still was false. The second logical option was padding R&S seperately leading to 0||R||0||S, which was eventually the correct format and lead to a working signature validation as expected.
